In [34]:
!apt-get update
!apt-get install ruby -y

Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:3 https://cli.github.com/packages stable InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:5 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:7 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:8 https://r2u.stat.illinois.edu/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Reading package lists... Done
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading packag

# DOCUMENTACION DE RUBY

Ruby es un lenguaje de programación interpretado y dinámico, conocido por su enfoque en la simplicidad y la productividad.


## 1. Mostrar en Consola y Comentarios

puts (put string) se usa para imprimir un valor en la consola, añadiendo automáticamente un salto de línea al final.

In [35]:
%%script ruby
puts "Bienvenido al mundo de Ruby!"

Bienvenido al mundo de Ruby!


print es similar, pero imprime el valor exacto sin añadir un salto de línea.

In [36]:
%%script ruby
print "Esto es "
print "una sola línea."

Esto es una sola línea.

In [37]:
%%script ruby
# Este es un comentario de una sola línea (el más usado).
puts "Los comentarios no se ejecutan."

=begin
Esto es un comentario
que abarca varias líneas.
No es tan común como el '#',
pero es útil para bloques largos.
=end

Los comentarios no se ejecutan.


## 2. Variables y Tipos de Datos Principales

Ruby utiliza tipado dinámico. Esto significa que no necesitas declarar el tipo de una variable; simplemente le asignas un valor y Ruby infiere el tipo.

In [38]:
%%script ruby
una_cadena = "Esto es una variable"
puts una_cadena

una_cadena = "Puedo cambiar mi valor fácilmente"
puts una_cadena

Esto es una variable
Puedo cambiar mi valor fácilmente


Dado que todo es un objeto, puedes usar el método .class en cualquier variable para descubrir su tipo (o clase).

In [39]:
%%script ruby
una_cadena = "Puedo cambiar mi valor fácilmente"
puts una_cadena.class # Imprime: String

String


##### Integer (Números Enteros)

In [40]:
%%script ruby
edad = 30
edad = edad + 1
puts edad
puts edad * 2
puts edad
puts edad.class

31
62
31
Integer


##### Float (Números Decimales)

In [41]:
%%script ruby
precio = 19.99
puts precio
puts precio.class

19.99
Float


Al ser objetos, los números también tienen métodos. Los métodos que terminan en ? suelen devolver un booleano.

In [42]:
%%script ruby
puts 5.odd? # Imprime: true (¿es impar?)

true


##### Booleanos (Verdadero/Falso)

In [43]:
%%script ruby
es_activo = true
esta_lloviendo = false
puts es_activo
puts es_activo.class
puts esta_lloviendo.class

true
TrueClass
FalseClass


##### Nil (Ausencia de Valor)

nil es el equivalente de Ruby a null o None. Representa "nada" y también es un objeto.

In [44]:
%%script ruby
sin_valor = nil
puts sin_valor       # No imprime nada, solo el salto de línea de puts
puts sin_valor.class # Imprime: NilClass


NilClass


##### Interpolación de Cadenas

Para insertar el valor de una variable dentro de una cadena, debes usar comillas dobles ("") y la sintaxis #{...}.

In [45]:
%%script ruby
edad = 30
edad = edad + 1
puts edad
puts edad * 2
puts edad
puts edad.class

es_activo = true
esta_lloviendo = false
puts es_activo
puts es_activo.class
puts esta_lloviendo.class

puts "La edad del usuario es #{edad} y su estado es #{es_activo}."
puts "El próximo año, la edad será #{edad + 1}." # Se puede ejecutar código dentro

31
62
31
Integer
true
TrueClass
FalseClass
La edad del usuario es 31 y su estado es true.
El próximo año, la edad será 32.


## 3. Constantes

Las constantes se definen comenzando con una letra mayúscula. Por convención, se usa SNAKE_CASE en mayúsculas.

In [46]:
%%script ruby
PI = 3.1415
puts PI
GRAVEDAD = 9.8
puts GRAVEDAD

3.1415
9.8


Aunque Ruby permite reasignar una constante, generará una advertencia. (Ej: PI = 3.14 mostraría un aviso).

## 4. Colecciones (Estructuras de Datos)

##### Array (Listas)

Los Arrays en Ruby son colecciones ordenadas. Pueden cambiar de tamaño (son dinámicos) y contener diferentes tipos de datos.

In [47]:
%%script ruby
mi_lista = ["Manzana", 100, true, 45.5]
puts mi_lista[0] # Imprime: Manzana (acceso por índice base 0)
puts mi_lista[2] # Imprime: true

Manzana
true


Puedes añadir elementos al final de un Array usando el método push o el operador << (pala).

In [48]:
%%script ruby
mi_lista = ["Manzana", 100, true, 45.5]
puts mi_lista[0] # Imprime: Manzana (acceso por índice base 0)
puts mi_lista[2] # Imprime: true

mi_lista.push("Nuevo elemento")
mi_lista << "Otro más"

Manzana
true


Para ver una representación clara del contenido del array (para depuración), usa .inspect.

In [49]:
%%script ruby
mi_lista = ["Manzana", 100, true, 45.5]
puts mi_lista[0] # Imprime: Manzana (acceso por índice base 0)
puts mi_lista[2] # Imprime: true

mi_lista.push("Nuevo elemento")
mi_lista << "Otro más"

puts mi_lista.inspect

Manzana
true
["Manzana", 100, true, 45.5, "Nuevo elemento", "Otro más"]


##### Hash (Mapas o Diccionarios)

Un Hash es una colección de pares clave-valor. La sintaxis moderna prefiere usar Símbolos (ej: :nombre) como claves, que son más eficientes que las cadenas.

In [50]:
%%script ruby
usuario = {
  nombre: "Fran",
  edad: 21,
  es_admin: true
}

Accedes a los valores utilizando la clave (el Símbolo).

In [51]:
%%script ruby
usuario = {
  nombre: "Fran",
  edad: 21,
  es_admin: true
}

puts usuario[:nombre] # Imprime: fran

# Añadir un nuevo par clave-valor
usuario[:ciudad] = "Valencia"
puts usuario.inspect

Fran
{:nombre=>"Fran", :edad=>21, :es_admin=>true, :ciudad=>"Valencia"}


##### Set (Conjuntos Únicos)

Un Set es una colección de elementos únicos y desordenados. Requiere importar la librería set.

In [52]:
%%script ruby
require 'set'

numeros_repetidos = [1, 2, 3, 2, 1, 4, 5, 1]
numeros_unicos = numeros_repetidos.to_set

puts numeros_unicos.inspect # Imprime: #<Set: {1, 2, 3, 4, 5}>

#<Set: {1, 2, 3, 4, 5}>


## 5. Estructuras de Control y Bucles

##### Condicionales (if/elsif/else)

En Ruby, casi todos los bloques de código (condicionales, bucles, métodos, clases) deben cerrarse con la palabra clave`end.

In [53]:
%%script ruby
nota = 7

if nota == 10
  puts "Excelente"
elsif nota >= 7
  puts "Aprobado"
elsif nota >= 5
  puts "Suficiente"
else
  puts "Suspendido"
end

Aprobado


##### Iteradores (La forma "Ruby")

La forma más idiomática de recorrer colecciones en Ruby no es con un bucle for, sino con iteradores como .each.

Iterando sobre Arrays

In [54]:
%%script ruby
puts "\nRecorriendo la lista:"
frutas = ["Manzana", "Banana", "Naranja"]

frutas.each do |fruta|
  # |fruta| es la variable temporal para cada elemento
  puts "Me gusta la #{fruta}"
end


Recorriendo la lista:
Me gusta la Manzana
Me gusta la Banana
Me gusta la Naranja


Iterando sobre Hashes

In [55]:
%%script ruby
usuario = {
  nombre: "Fran",
  edad: 21,
  es_admin: true
}

puts usuario[:nombre] # Imprime: fran

# Añadir un nuevo par clave-valor
usuario[:ciudad] = "Valencia"
puts usuario.inspect

puts "\nRecorriendo el hash de usuario:"
usuario.each do |clave, valor|
  puts "  - #{clave}: #{valor}"
end

Fran
{:nombre=>"Fran", :edad=>21, :es_admin=>true, :ciudad=>"Valencia"}

Recorriendo el hash de usuario:
  - nombre: Fran
  - edad: 21
  - es_admin: true
  - ciudad: Valencia


##### Bucles Clásicos

Bucle for: Aunque menos común (se prefiere .each), existe. Usa un **Rango**, como (1..3) que incluye 1, 2 y 3.

In [56]:
%%script ruby
puts "\nBucle for:"
for i in 1..3 do
  puts "Número #{i}"
end


Bucle for:
Número 1
Número 2
Número 3


Bucle 'while'

In [57]:
%%script ruby
puts "\nBucle while:"
contador = 3
while contador > 0 do
  puts "Tiempo restante: #{contador}"
  contador -= 1
end
puts "¡Despegue!"


Bucle while:
Tiempo restante: 3
Tiempo restante: 2
Tiempo restante: 1
¡Despegue!


## 6. Métodos (Funciones)

Los métodos (similares a las funciones) se definen con def. Los paréntesis para los parámetros y al llamar al método son a menudo opcionales.

In [58]:
%%script ruby
def saludar(nombre)
  puts "Hola, #{nombre}!"
end

# Se pueden llamar con o sin paréntesis
saludar("Carlos")
saludar "Maria"

Hola, Carlos!
Hola, Maria!


##### Retorno Implícito

Ruby utiliza el retorno implícito: el valor de la última expresión ejecutada en un método se devuelve automáticamente. La palabra clave return es opcional.

In [59]:
%%script ruby
def sumar(a, b)
  a + b
end

resultado = sumar(5, 3)
puts "El resultado de la suma es: #{resultado}" # Imprime 8

El resultado de la suma es: 8
